paymentapi/src/main/java/fi/hel/verkkokauppa/payment/service/refund/PaytrailRefundPaymentService.java

```diff
@@ -0,0 +1,343 @@
+    }
+
+
+    private RefundPayment selectPayableRefundPayment(List<RefundPayment> payments) {
+        RefundPayment payablePayment = null;
+
+        if (payments != null)
+            for (RefundPayment payment : payments) {
+                // in an unpayable state
+                if (Objects.equals(payment.getStatus(), RefundPaymentStatus.PAID_ONLINE) || Objects.equals(payment.getStatus(), RefundPaymentStatus.CANCELLED)) {
+                    continue;
+                }
+
+                // an earlier selected refund payment is newer
+                if (payablePayment != null && payablePayment.getTimestamp().compareTo(payment.getTimestamp()) > 0) {
+                    continue;
+                }
+
+                payablePayment = payment;
+            }
+
+        return payablePayment;
+    }
+
+    private RefundPayment selectPaidRefundPayment(List<RefundPayment> payments) {
+        if (payments != null) {
+            for (RefundPayment payment : payments) {
+                if (Objects.equals(payment.getStatus(), RefundPaymentStatus.PAID_ONLINE)) {
+                    return payment;
+                }
+            }
+        }
+
+        return null;
+    }
+
+    public RefundPayment getRefund(String refundId) {
+        Optional<RefundPayment> payment = refundPaymentRepository.findById(refundId);
+
+        if (!payment.isPresent()) {
+            log.debug("refund not found, refundId: " + refundId);
+            Error error = new Error("refund-not-found-from-backend", "refund with refund id [" + refundId + "] not found from backend");
+            throw new CommonApiException(HttpStatus.NOT_FOUND, error);
+        }
+
+        return payment.get();
+    }
+
+    public RefundPayment createRefundToPaytrailAndCreateRefundPayment(RefundRequestDataDto dto) {
+        RefundDto refundDto = dto.getRefund().getRefund();
+        String refundId = refundDto.getRefundId();
+
+        isValidRefundStatusToCreateRefundPayment(
+                refundId,
+                refundDto.getStatus()
+        );
+
+        isValidUserToCreateRefundPayment(
+                refundId,
+                refundDto.getUser()
+        );
+        String orderType = getOrderType(dto.getOrder().getOrder());
+
+        PaymentDto paymentDto = dto.getPayment();
+        // Get merchantId from first refund item
+        RefundItemDto refundItemDto = dto.getRefund().getItems().get(0);
+        PaytrailPaymentContext context = createPaytrailRefundContext(refundDto, refundItemDto, paymentDto);
+
+        String paymentTransactionId = paymentDto.getPaytrailTransactionId();
+        String refundPaymentId = RefundUtil.generateRefundPaymentId(refundId);
+
+        PaytrailRefundResponse refundResponse = paytrailPaymentClient.createRefund(
+                context,
+                refundPaymentId,
+                paymentTransactionId,
+                refundDto
+        );
+
+        if (!refundResponse.isValid()) {
```

Comment:
This if is unnecessary, beacause **paytrailPaymentClient.createRefund(..)** will throw CommonApiException if **!isValid**.